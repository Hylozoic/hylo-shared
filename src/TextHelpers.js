import { convert as convertHtmlToText } from 'html-to-text'
import { isURL } from 'validator'
import { marked } from 'marked'
import merge from 'lodash/fp/merge'
import { DateTime } from 'luxon'
import prettyDate from 'pretty-date'
import truncHTML from 'trunc-html'
import truncText from 'trunc-text'

// Sanitization options
export function insaneOptions (providedInsaneOptions) {
  return merge(
    {
      allowedTags: providedInsaneOptions?.allowedTags || [
        'a', 'br', 'em', 'p', 's', 'strong',
        'li', 'ol', 'ul',
        'div', 'iframe', 'mark', 'span',
        'blockquote', 'code', 'hr', 'pre',
        'h1', 'h2', 'h3', 'h4', 'h5', 'h6'
      ],
      allowedAttributes: providedInsaneOptions?.allowedAttributes || {
        a: [
          'class', 'target', 'href',
          'data-type', 'data-id', 'data-label',
          'data-user-id', 'data-entity-type', 'data-search'
        ],
        span: [
          'class', 'target', 'href',
          'data-type', 'data-id', 'data-label',
          'data-user-id', 'data-entity-type', 'data-search'
        ],
        code: [
          'class'
        ],
        iframe: [
          'src', 'title', 'frameborder', 'height', 'width',
          'allow', 'allowfullscreen'
        ],
        div: [
          'class'
        ]
      }
    },
    providedInsaneOptions
  )
}

export const truncateHTML = (html, truncateLength, providedInsaneOptions = {}) => {
  const options = {
    sanitizer: insaneOptions(providedInsaneOptions)
  }

  return truncHTML(html, truncateLength, options).html
}

export function presentHTMLToText (html, options = {}) {
  if (!html) return ''

  const { truncate: truncateLength, providedConvertHtmlToTextOptions = {} } = options
  const convertHtmlToTextOptions = merge(
    {
      selectors: [
        {
          selector: 'a',
          options: {
            ignoreHref: true
          }
        }
      ]
    },
    providedConvertHtmlToTextOptions
  )

  let processedText = convertHtmlToText(html, convertHtmlToTextOptions)

  if (truncateLength) {
    processedText = truncateText(processedText, truncateLength)
  }

  return processedText
}

export const truncateText = (text, length) => {
  return truncText(text, length)
}

export function textLengthHTML (htmlOrText, options) {
  return presentHTMLToText(htmlOrText, options).length
}

export const markdown = (text, options = {}) => {
  if (options.disableAutolinking) {
    marked.use({
      tokenizer: {
        url (src) {
          // NOTE: having nothing here disables gfm autolinks:
          // https://github.com/markedjs/marked/issues/882#issuecomment-781585009
          // New option coming in later version: https://github.com/sourcegraph/sourcegraph/pull/42203/files
        }
      }
    })
  }

  return marked.parse(text || '', { gfm: true, breaks: true })
}

// HTML Generation Helpers

export const mentionHTML = mentioned => (
  `<span data-type="mention" class="mention" data-id="${mentioned.id}" data-label="${mentioned.name}">${mentioned.name}</span>`
)

export const topicHTML = topicName => (
  `<span data-type="topic" class="topic" data-id="${topicName}" data-label="#${topicName}">${topicName}</span>`
)

// URL helpers

export const sanitizeURL = url => {
  if (!url) return null
  if (isURL(url, { require_protocol: true })) return url
  if (isURL(`https://${url}`)) return `https://${url}`
  return null
}

// Date string related

export function humanDate (date, short) {
  const isString = typeof date === 'string'
  const isValidDate = !isNaN(Number(date)) && Number(date) !== 0
  let ret = date && (isString || isValidDate)
    ? prettyDate.format(isString ? new Date(date) : date)
    : ''

  if (short) {
    ret = ret.replace(' ago', '')
  } else {
    // this workaround prevents a "React attempted to use reuse markup" error
    // which happens if the timestamp is less than 1 minute ago, because the
    // server renders "N seconds ago", but by the time React is loaded on the
    // client side, it's "N+1 seconds ago"
    const match = ret.match(/(\d+) seconds? ago/)
    if (match) {
      if (Number(match[1]) >= 50) return '1m ago'
      return 'just now'
    }
  }

  return ret.replace(/ minutes?/, 'm')
    .replace(/ hours?/, 'h')
    .replace(/ days?/, 'd')
    .replace(/ weeks?/, 'w')
    .replace(/ month(s?)/, ' mo$1')
}

export const formatDatePair = (locale = 'en', startTime, endTime, returnAsObj) => {
  const dateWithTime = { ...DateTime.DATE_MED_WITH_WEEKDAY, hour: 'numeric', minute: 'numeric' }
  const dateWithTimeAndOffset = { ...DateTime.DATE_MED_WITH_WEEKDAY, hour: 'numeric', minute: 'numeric', timeZoneName: 'short' }
  const dateNoYearWithTime = { weekday: 'short', month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric' }
  const dateNoYearWithTimeAndOffset = { weekday: 'short', month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric', timeZoneName: 'short' }
  const start = startTime ? DateTime.fromISO(startTime, { locale }) : null
  const end = endTime ? DateTime.fromISO(endTime, { locale }) : null

  const now = DateTime.local({ locale })
  const isThisYear = start.year === now.year && end.year === now.year

  let to = ''
  let from = ''

  if (isThisYear) {
    from = endTime ? start.toLocaleString(dateNoYearWithTime) : start.toLocaleString(dateNoYearWithTimeAndOffset)
  } else {
    from = endTime ? start.toLocaleString(dateWithTime) : start.toLocaleString(dateWithTimeAndOffset)
  }
  if (endTime) {
    if (end.year !== start.year && !isThisYear) {
      to = end.toLocaleString(dateWithTimeAndOffset)
    } else if (end.month !== start.month ||
               end.day !== start.day ||
               end <= now) {
      to = end.toLocaleString(dateNoYearWithTimeAndOffset)
    } else {
      to = end.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET)
    }
    to = returnAsObj ? to : ' - ' + to
  }
  return returnAsObj ? { from, to } : from + to
}

export function isDateInTheFuture (date) {
  return typeof date === 'number' ? DateTime.fromMillis(date) > DateTime.now() : DateTime.fromISO(date) > DateTime.now()
}
