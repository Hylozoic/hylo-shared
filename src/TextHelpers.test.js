import * as TextHelpers from '../src/TextHelpers'
import { DateTime } from 'luxon'

describe('presentHTMLToText', () => {
  it("shouldn't include text of href on links", () => {
    expect(TextHelpers.presentHTMLToText("<a href='/any/url'>Text</a> more text")).toBe('Text more text')
  })
})

describe('truncateText', () => {
  it('has an ellipses after truncation', () => {
    expect(TextHelpers.truncateText('<I mean it> test test test', 8)).toBe('<I mean â€¦')
  })

  it('does not have an ellipses if there was no truncation', () => {
    expect(TextHelpers.truncateText('<I mean it> test test test', 100)).toBe('<I mean it> test test test')
  })
})

describe('textLengthHTML', () => {
  it('should return lenght of plain text version of the html', () => {
    expect(TextHelpers.textLengthHTML('<strong>test</strong> <a href="">a link</a>')).toBe(11)
  })
})

describe('markdown', () => {
  it('converts to markdown', () => {
    expect(TextHelpers.markdown('*strong* **italic**')).toBe('<p><em>strong</em> <strong>italic</strong></p>\n')
  })
  it('converts to markdown with autolinking (default)', () => {
    expect(TextHelpers.markdown('https://www.hylo.com')).toBe('<p><a href="https://www.hylo.com">https://www.hylo.com</a></p>\n')
  })
  it('converts to markdown with disableAutolinking', () => {
    expect(TextHelpers.markdown('https://www.hylo.com', { disableAutolinking: true })).toBe('<p>https://www.hylo.com</p>\n')
  })
  it('converts to markdown in paragraphs', () => {
    expect(TextHelpers.markdown('asdw\n\n\nasdf')).toBe('<p>asdw</p>\n<p>asdf</p>\n')
  })
})

describe('formatDatePair', () => {
  it('displays differences of dates', () => {
    const d1 = DateTime.local(2019, 2, 4, 11, 41)
    const d2 = DateTime.local(2019, 2, 4, 14, 41)
    const d3 = DateTime.local(2019, 2, 5, 14, 41)
    const d4 = DateTime.local(2019, 3, 5, 14, 41)
    const d5 = DateTime.local(2050, 2, 5, 14, 41)
    const d6 = DateTime.local(DateTime.now().year, 2, 5, 14, 41)
    const d7 = DateTime.local(DateTime.now().year, 3, 5, 14, 41)

    expect(TextHelpers.formatDatePair(d1)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d2)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d3)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d4)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d5)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d6)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d6, d5)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d6, d7)).toMatchSnapshot()
  })
})

describe('sanitizeURL', () => {
  it('adds "https" when no protocol specified', () => {
    expect(TextHelpers.sanitizeURL('www.hylo.com')).toBe('https://www.hylo.com')
  })
})
