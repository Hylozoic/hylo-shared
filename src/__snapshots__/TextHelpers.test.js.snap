// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDatePair displays differences of dates 1`] = `"Mon, Feb 4, 2019, 11:41 AM CST"`;

exports[`formatDatePair displays differences of dates 2`] = `"Mon, Feb 4, 2019, 11:41 AM - Mon, Feb 4, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 3`] = `"Mon, Feb 4, 2019, 11:41 AM - Tue, Feb 5, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 4`] = `"Mon, Feb 4, 2019, 11:41 AM - Tue, Mar 5, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 5`] = `"Mon, Feb 4, 2019, 11:41 AM - Sat, Feb 5, 2050, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 6`] = `"Mon, Feb 4, 2019, 11:41 AM - Sun, Feb 5, 2023, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 7`] = `"Sun, Feb 5, 2023, 2:41 PM - Sat, Feb 5, 2050, 2:41 PM CST"`;

exports[`formatDatePair displays differences of dates 8`] = `"Sun, Feb 5, 2:41 PM - Sun, Mar 5, 2:41 PM CST"`;
